// claude-prompts-server/dist/text-references/conversation.js
// @lifecycle canonical - Tracks conversation references for prompts.
/**
 * Conversation Management Module
 * Maintains lightweight conversation history for tooling and diagnostics.
 */
export class ConversationManager {
    constructor(logger, maxHistorySize = 100) {
        this.conversationHistory = [];
        this.logger = logger;
        this.maxHistorySize = maxHistorySize;
    }
    /**
     * Append a conversation item with bounded history management.
     */
    addToConversationHistory(item) {
        this.conversationHistory.push(item);
        if (this.conversationHistory.length > this.maxHistorySize) {
            this.conversationHistory.splice(0, this.conversationHistory.length - this.maxHistorySize);
            this.logger.debug(`Trimmed conversation history to ${this.maxHistorySize} entries to prevent memory leaks`);
        }
    }
    /**
     * Locate the most recent non-template user message for template context.
     */
    getPreviousMessage() {
        for (let i = this.conversationHistory.length - 1; i >= 0; i--) {
            const historyItem = this.conversationHistory[i];
            if (historyItem.role === 'user' && !historyItem.isProcessedTemplate) {
                this.logger.debug(`Found previous user message for context: ${historyItem.content.substring(0, 50)}...`);
                return historyItem.content;
            }
        }
        return '[Please check previous messages in the conversation for context]';
    }
    /**
     * Return a shallow copy of the recorded history.
     */
    getConversationHistory() {
        return [...this.conversationHistory];
    }
    /**
     * Short-hand helper for UIs needing limited history snapshots.
     */
    getRecentMessages(count = 5) {
        return this.conversationHistory.slice(-count);
    }
    /**
     * Clear all stored conversation entries.
     */
    clearHistory() {
        this.conversationHistory = [];
        this.logger.info('Conversation history cleared');
    }
    /**
     * Provide high-level stats for diagnostics.
     */
    getConversationStats() {
        const userMessages = this.conversationHistory.filter((item) => item.role === 'user').length;
        const assistantMessages = this.conversationHistory.filter((item) => item.role === 'assistant').length;
        const processedTemplates = this.conversationHistory.filter((item) => item.isProcessedTemplate).length;
        const timestamps = this.conversationHistory.map((item) => item.timestamp);
        const oldestMessage = timestamps.length > 0 ? Math.min(...timestamps) : undefined;
        const newestMessage = timestamps.length > 0 ? Math.max(...timestamps) : undefined;
        return {
            totalMessages: this.conversationHistory.length,
            userMessages,
            assistantMessages,
            processedTemplates,
            oldestMessage,
            newestMessage,
        };
    }
}
export function createConversationManager(logger, maxHistorySize) {
    return new ConversationManager(logger, maxHistorySize);
}
