// claude-prompts-server/dist/frameworks/types/methodology-types.js
// @lifecycle canonical - Core framework and methodology type definitions.
/**
 * Methodology Guide Type Definitions
 *
 * Contains all types related to methodology guides, framework definitions,
 * and methodology-specific interfaces. This consolidates types from multiple
 * sources to eliminate duplication.
 */
/**
 * Base class for methodology guides
 * Provides common functionality for all methodology implementations
 */
export class BaseMethodologyGuide {
    /**
     * Helper method to extract combined text from prompt
     */
    getCombinedText(prompt) {
        return [prompt.systemMessage || '', prompt.userMessageTemplate || '', prompt.description || '']
            .filter((text) => text.trim())
            .join(' ');
    }
    /**
     * Helper method to create enhancement metadata
     */
    createEnhancementMetadata(confidence, reason) {
        return {
            methodology: this.methodology,
            confidence,
            applicabilityReason: reason,
            appliedAt: new Date(),
        };
    }
}
