// claude-prompts-server/dist/runtime/options.js
// @lifecycle canonical - Parses runtime CLI options and flags.
/**
 * Runtime launch option parsing shared between the CLI entrypoint and the runtime application.
 * Centralizes CLI flag handling so verbose/quiet/test detection stays in sync.
 */
const TEST_ARG_HINTS = ['test', 'jest', 'mocha'];
/**
 * Determine whether the current process is executing under a test harness or CI.
 */
export function detectRuntimeTestEnvironment(fullArgv = process.argv, args = process.argv.slice(2)) {
    return (process.env.NODE_ENV === 'test' ||
        args.includes('--suppress-debug') ||
        args.includes('--test-mode') ||
        process.env.GITHUB_ACTIONS === 'true' ||
        process.env.CI === 'true' ||
        fullArgv.some((arg) => TEST_ARG_HINTS.some((hint) => arg.includes(hint))) ||
        (fullArgv[1] ?? '').includes('tests/scripts/'));
}
/**
 * Parse runtime launch options from CLI arguments.
 */
export function resolveRuntimeLaunchOptions(args = process.argv.slice(2), fullArgv = process.argv) {
    return {
        args,
        verbose: args.includes('--verbose') || args.includes('--debug-startup'),
        quiet: args.includes('--quiet'),
        startupTest: args.includes('--startup-test'),
        testEnvironment: detectRuntimeTestEnvironment(fullArgv, args),
    };
}
