// claude-prompts-server/dist/execution/parsers/index.js
// @lifecycle canonical - Barrel exports for execution parsers.
/**
 * Execution Parsers Export Module
 *
 * Centralizes all parsing infrastructure exports including:
 * - Unified Command Parser with multi-strategy parsing
 * - Argument Processing Pipeline with validation and enrichment
 * - Compatibility Wrapper for backward compatibility
 */
// Core parsing infrastructure
export { UnifiedCommandParser, createUnifiedCommandParser, } from './command-parser.js';
export { ArgumentParser, createArgumentParser, } from './argument-parser.js';
// Context resolution system
export { ContextResolver, createContextResolver, } from '../context/context-resolver.js';
import { createArgumentParser } from './argument-parser.js';
import { createUnifiedCommandParser } from './command-parser.js';
import { createContextResolver } from '../context/context-resolver.js';
/**
 * Factory function to create complete parsing system
 *
 * Creates a fully configured parsing system with:
 * - Unified command parser with multi-strategy support
 * - Argument processor with validation and type coercion
 * - Context resolver with intelligent fallbacks
 *
 * @param logger Logger instance for system-wide logging
 * @returns Complete parsing system ready for use
 */
export function createParsingSystem(logger) {
    const commandParser = createUnifiedCommandParser(logger);
    const argumentParser = createArgumentParser(logger);
    const contextResolver = createContextResolver(logger);
    logger.info('Parsing system initialized successfully');
    logger.info('- Unified command parser with multi-strategy support');
    logger.info('- Argument parser with validation pipeline');
    logger.info('- Context resolver with intelligent fallbacks');
    return {
        commandParser,
        argumentParser,
        contextResolver,
    };
}
