// claude-prompts-server/dist/utils/yaml/yaml-parser.js
// @lifecycle canonical - Core YAML parsing with comprehensive error handling
/**
 * YAML Parser Utilities
 *
 * Provides type-safe YAML parsing with detailed error handling.
 * Designed to be reusable across methodologies and future prompt YAML support.
 */
import yaml from 'js-yaml';
/**
 * Parse YAML content with comprehensive error handling
 *
 * Returns a result object with success/error info rather than throwing.
 * Use this when you want to handle parsing failures gracefully.
 *
 * @param content - YAML string to parse
 * @param options - Parsing options
 * @returns Result object with parsed data or error details
 *
 * @example
 * ```typescript
 * const result = parseYaml<Config>(yamlString, { filename: 'config.yaml' });
 * if (result.success) {
 *   console.log(result.data);
 * } else {
 *   console.error(`Error at line ${result.error?.line}: ${result.error?.message}`);
 * }
 * ```
 */
export function parseYaml(content, options) {
    const warnings = [];
    try {
        const data = yaml.load(content, {
            schema: options?.schema ?? yaml.DEFAULT_SCHEMA,
            filename: options?.filename,
            onWarning: (warning) => {
                warnings.push(warning.message);
                options?.onWarning?.(warning);
            },
        });
        return {
            success: true,
            data,
            warnings: warnings.length > 0 ? warnings : undefined,
        };
    }
    catch (error) {
        if (error instanceof yaml.YAMLException) {
            return {
                success: false,
                error: {
                    message: error.message,
                    filename: options?.filename,
                    line: error.mark?.line,
                    column: error.mark?.column,
                    snippet: error.mark?.snippet,
                    cause: error,
                },
                warnings: warnings.length > 0 ? warnings : undefined,
            };
        }
        // Handle unexpected errors
        return {
            success: false,
            error: {
                message: error instanceof Error ? error.message : String(error),
                filename: options?.filename,
                cause: error instanceof Error ? error : undefined,
            },
            warnings: warnings.length > 0 ? warnings : undefined,
        };
    }
}
/**
 * Parse YAML content, throwing on error
 *
 * Use when parsing failure should halt execution.
 * Provides detailed error messages for debugging.
 *
 * @param content - YAML string to parse
 * @param options - Parsing options
 * @returns Parsed data
 * @throws Error with detailed message on parse failure
 *
 * @example
 * ```typescript
 * try {
 *   const config = parseYamlOrThrow<Config>(yamlString, { filename: 'config.yaml' });
 * } catch (error) {
 *   // Error message includes filename, line, column
 * }
 * ```
 */
export function parseYamlOrThrow(content, options) {
    const result = parseYaml(content, options);
    if (!result.success) {
        const error = result.error;
        const location = error.line !== undefined ? ` at line ${error.line + 1}` : '';
        const column = error.column !== undefined ? `:${error.column}` : '';
        const file = error.filename ? ` in ${error.filename}` : '';
        throw new Error(`YAML parse error${file}${location}${column}: ${error.message}`);
    }
    return result.data;
}
/**
 * Serialize data to YAML string
 *
 * @param data - Data to serialize
 * @param options - Serialization options
 * @returns YAML string representation
 */
export function serializeYaml(data, options) {
    return yaml.dump(data, {
        indent: options?.indent ?? 2,
        lineWidth: options?.lineWidth ?? 80,
        noRefs: options?.noRefs ?? true,
        sortKeys: options?.sortKeys ?? false,
    });
}
/**
 * Format a YAML parse error for logging
 *
 * @param error - Parse error to format
 * @returns Formatted error string
 */
export function formatYamlError(error) {
    const parts = [];
    if (error.filename) {
        parts.push(`File: ${error.filename}`);
    }
    if (error.line !== undefined) {
        const location = error.column !== undefined ? `${error.line + 1}:${error.column}` : `line ${error.line + 1}`;
        parts.push(`Location: ${location}`);
    }
    parts.push(`Error: ${error.message}`);
    if (error.snippet) {
        parts.push(`\nContext:\n${error.snippet}`);
    }
    return parts.join('\n');
}
