// claude-prompts-server/dist/mcp-tools/prompt-manager/search/filter-parser.js
// @lifecycle canonical - Parses filter expressions for prompt searches.
/**
 * Intelligent filter parsing for prompt discovery
 */
import { validateFilterSyntax } from '../utils/validation.js';
/**
 * Filter parsing engine for intelligent prompt discovery
 */
export class FilterParser {
    constructor(logger) {
        this.logger = logger;
    }
    /**
     * Parse intelligent filters for list operation
     */
    parseIntelligentFilters(filterText) {
        const filters = {};
        if (!filterText)
            return filters;
        try {
            // Validate filter syntax
            validateFilterSyntax(filterText);
            // Parse various filter patterns
            this.parseTypeFilter(filterText, filters);
            this.parseCategoryFilter(filterText, filters);
            this.parseIntentFilter(filterText, filters);
            this.parseExecutionFilter(filterText, filters);
            this.parseGatesFilter(filterText, filters);
            // Extract remaining text as search term
            const cleanedText = this.extractTextFilter(filterText);
            if (cleanedText) {
                filters.text = cleanedText;
            }
            this.logger.info(`Parsed filters for "${filterText}":`, filters);
        }
        catch (error) {
            this.logger.warn(`Filter parsing error: ${error instanceof Error ? error.message : String(error)}`);
            // Return text-only filter as fallback
            filters.text = filterText;
        }
        return filters;
    }
    /**
     * Parse type filter (type:prompt, type:template, type:chain)
     */
    parseTypeFilter(filterText, filters) {
        const typeMatch = filterText.match(/type:(\w+)/i);
        if (typeMatch) {
            const type = typeMatch[1].toLowerCase();
            if (['prompt', 'template', 'chain'].includes(type)) {
                filters.type = type;
            }
        }
    }
    /**
     * Parse category filter (category:code, category:analysis)
     */
    parseCategoryFilter(filterText, filters) {
        const categoryMatch = filterText.match(/category:([a-z-_]+)/i);
        if (categoryMatch) {
            filters.category = categoryMatch[1].toLowerCase();
        }
    }
    /**
     * Parse intent filter (intent:debugging, intent:analysis)
     */
    parseIntentFilter(filterText, filters) {
        const intentMatch = filterText.match(/intent:([a-z-_\s]+)/i);
        if (intentMatch) {
            filters.intent = intentMatch[1].trim().toLowerCase();
        }
    }
    /**
     * Parse execution requirement filter (execution:required, execution:optional)
     */
    parseExecutionFilter(filterText, filters) {
        if (filterText.includes('execution:required')) {
            filters.execution = true;
        }
        else if (filterText.includes('execution:optional')) {
            filters.execution = false;
        }
    }
    /**
     * Parse gates filter (gates:yes, gates:no)
     */
    parseGatesFilter(filterText, filters) {
        if (filterText.includes('gates:yes')) {
            filters.gates = true;
        }
        else if (filterText.includes('gates:no')) {
            filters.gates = false;
        }
    }
    /**
     * Extract text search terms after removing filter syntax
     */
    extractTextFilter(filterText) {
        const cleanedText = filterText
            .replace(/type:\w+/gi, '')
            .replace(/category:[a-z-_]+/gi, '')
            .replace(/intent:[a-z-_\s]+/gi, '')
            .replace(/confidence:[<>]?\d+(?:-\d+)?/g, '')
            .replace(/execution:(required|optional)/gi, '')
            .replace(/gates:(yes|no)/gi, '')
            .replace(/\s+/g, ' ')
            .trim();
        return cleanedText;
    }
    /**
     * Build filter description for display
     */
    buildFilterDescription(filters) {
        const descriptions = [];
        if (filters.type) {
            descriptions.push(`**Type**: ${filters.type}`);
        }
        if (filters.category) {
            descriptions.push(`**Category**: ${filters.category}`);
        }
        if (filters.intent) {
            descriptions.push(`**Intent**: "${filters.intent}"`);
        }
        if (filters.text) {
            descriptions.push(`**Search**: "${filters.text}"`);
        }
        if (filters.execution !== undefined) {
            descriptions.push(`**Execution**: ${filters.execution ? 'Required' : 'Optional'}`);
        }
        if (filters.gates !== undefined) {
            descriptions.push(`**Quality Gates**: ${filters.gates ? 'Required' : 'None'}`);
        }
        return descriptions;
    }
    /**
     * Generate filter examples for help
     */
    getFilterExamples() {
        return [
            'type:template - Show only template prompts',
            'category:analysis - Show prompts in analysis category',
            'intent:debugging - Find prompts for debugging tasks',
            'execution:required - Show prompts that require execution',
            'gates:yes - Show prompts with quality gates',
            'react component - Text search for "react component"',
            'type:chain category:development - Combined filters',
        ];
    }
    /**
     * Suggest filters based on common patterns
     */
    suggestFilters(searchText) {
        const suggestions = [];
        const text = searchText.toLowerCase();
        // Suggest type filters based on common terms
        if (text.includes('template') || text.includes('variable')) {
            suggestions.push('type:template');
        }
        if (text.includes('chain') || text.includes('multi') || text.includes('step')) {
            suggestions.push('type:chain');
        }
        if (text.includes('simple') || text.includes('basic')) {
            suggestions.push('type:prompt');
        }
        // Suggest category filters based on common terms
        if (text.includes('code') || text.includes('develop') || text.includes('program')) {
            suggestions.push('category:development');
        }
        if (text.includes('analy') || text.includes('review') || text.includes('examine')) {
            suggestions.push('category:analysis');
        }
        if (text.includes('research') || text.includes('investigate')) {
            suggestions.push('category:research');
        }
        // Suggest intent filters based on common tasks
        if (text.includes('debug') || text.includes('fix') || text.includes('error')) {
            suggestions.push('intent:debugging');
        }
        if (text.includes('test') || text.includes('validate')) {
            suggestions.push('intent:testing');
        }
        if (text.includes('optimize') || text.includes('improve')) {
            suggestions.push('intent:optimization');
        }
        return suggestions.slice(0, 3); // Limit to 3 suggestions
    }
    /**
     * Validate filter combination
     */
    validateFilterCombination(filters) {
        const warnings = [];
        // Check for conflicting filters
        if (filters.type === 'prompt' && filters.gates === true) {
            warnings.push('Basic prompts typically do not have quality gates');
        }
        if (filters.execution === false && filters.type === 'chain') {
            warnings.push('Chain prompts typically require execution');
        }
        if (filters.intent && filters.category) {
            // Check if intent and category are compatible
            const categoryIntentMap = {
                development: ['debugging', 'testing', 'optimization'],
                analysis: ['research', 'investigation', 'review'],
                content: ['writing', 'editing', 'creation'],
            };
            const compatibleIntents = categoryIntentMap[filters.category] || [];
            if (compatibleIntents.length > 0 &&
                !compatibleIntents.some((intent) => filters.intent?.includes(intent))) {
                warnings.push(`Intent "${filters.intent}" may not match category "${filters.category}"`);
            }
        }
        return {
            valid: warnings.length === 0,
            warnings,
        };
    }
}
