// claude-prompts-server/dist/gates/services/compositional-gate-service.js
const DEFAULT_CONFIG = {
    enabled: true,
};
/**
 * Compositional Gate Service - Template rendering only (no server-side validation)
 *
 * Simplified to use GateGuidanceRenderer directly, removing the unnecessary
 * GateInstructionInjector abstraction layer.
 */
export class CompositionalGateService {
    constructor(logger, gateGuidanceRenderer, config) {
        this.serviceType = 'compositional';
        this.logger = logger;
        this.gateGuidanceRenderer = gateGuidanceRenderer;
        this.config = { ...DEFAULT_CONFIG, ...config };
    }
    async enhancePrompt(prompt, gateIds, context) {
        if (!this.config.enabled || gateIds.length === 0) {
            return {
                enhancedPrompt: prompt,
                gateInstructionsInjected: false,
                injectedGateIds: [],
            };
        }
        this.logger.debug('[CompositionalGateService] Rendering gate guidance', {
            promptId: prompt.id,
            gateCount: gateIds.length,
        });
        try {
            const guidance = await this.gateGuidanceRenderer.renderGuidance(gateIds, {
                framework: context.framework,
                category: context.category ?? prompt.category,
                promptId: context.promptId ?? prompt.id,
                explicitGateIds: context.explicitGateIds,
            });
            if (!guidance || guidance.trim().length === 0) {
                this.logger.debug('[CompositionalGateService] No guidance produced', {
                    promptId: prompt.id,
                });
                return {
                    enhancedPrompt: prompt,
                    gateInstructionsInjected: false,
                    injectedGateIds: [],
                };
            }
            const template = prompt.userMessageTemplate ?? '';
            const enhancedTemplate = template.length > 0 ? `${template}\n\n${guidance}` : guidance;
            this.logger.debug('[CompositionalGateService] Gate guidance injected', {
                promptId: prompt.id,
                gateCount: gateIds.length,
                guidanceLength: guidance.length,
            });
            return {
                enhancedPrompt: {
                    ...prompt,
                    userMessageTemplate: enhancedTemplate,
                },
                gateInstructionsInjected: true,
                injectedGateIds: gateIds,
                instructionLength: guidance.length,
            };
        }
        catch (error) {
            this.logger.error('[CompositionalGateService] Failed to render gate guidance', {
                error,
                promptId: prompt.id,
                gateIds,
            });
            return {
                enhancedPrompt: prompt,
                gateInstructionsInjected: false,
                injectedGateIds: [],
            };
        }
    }
    supportsValidation() {
        return false;
    }
    updateConfig(config) {
        this.config = { ...this.config, ...config };
    }
}
