// claude-prompts-server/dist/execution/pipeline/stages/00-dependency-injection-stage.js
// @lifecycle canonical - Initializes shared dependencies for downstream stages.
import { BasePipelineStage } from '../stage.js';
import { GateEnforcementAuthority } from '../decisions/index.js';
/**
 * Canonical Pipeline Stage 0.2: Dependency Injection
 *
 * Records execution dependencies (framework state, gate registry, analytics)
 * directly on the ExecutionContext so downstream stages can access a single
 * source of truth without recreating wiring from PromptExecutionService.
 */
export class DependencyInjectionStage extends BasePipelineStage {
    constructor(temporaryGateRegistry, chainSessionManager, frameworkEnabledProvider, metricsProvider, pipelineVersion, logger) {
        super(logger);
        this.temporaryGateRegistry = temporaryGateRegistry;
        this.chainSessionManager = chainSessionManager;
        this.frameworkEnabledProvider = frameworkEnabledProvider;
        this.metricsProvider = metricsProvider;
        this.pipelineVersion = pipelineVersion;
        this.name = 'DependencyInjection';
    }
    async execute(context) {
        this.logEntry(context);
        const frameworkEnabled = this.frameworkEnabledProvider?.() ?? false;
        const analyticsService = this.metricsProvider?.();
        // Initialize gate enforcement authority for downstream stages
        context.gateEnforcement = new GateEnforcementAuthority(this.chainSessionManager, this.logger);
        context.metadata['pipelineDependencies'] = {
            frameworkEnabled,
            analyticsService,
            temporaryGateRegistry: this.temporaryGateRegistry,
            pipelineVersion: this.pipelineVersion,
        };
        if (!context.metadata['executionOptions']) {
            context.metadata['executionOptions'] = {
                llmValidation: context.hasLlmValidation(),
                options: context.mcpRequest.options,
            };
        }
        this.logExit({
            frameworkEnabled,
            analyticsAttached: Boolean(analyticsService),
            gateEnforcementInitialized: true,
        });
    }
}
