// claude-prompts-server/dist/frameworks/methodology/index.js
// @lifecycle canonical - Barrel exports for methodology registry and guides.
/**
 * Methodology System Index
 *
 * Centralized exports for the reorganized methodology system.
 * Uses YAML-based loading exclusively via RuntimeMethodologyLoader.
 */
// Export methodology registry
export { MethodologyRegistry, createMethodologyRegistry } from './registry.js';
// Export methodology interfaces
export * from './interfaces.js';
// Export data-driven methodology guide system
export { GenericMethodologyGuide, createGenericGuide } from './generic-methodology-guide.js';
// Export runtime YAML loader (canonical source for methodology loading)
export { RuntimeMethodologyLoader, createRuntimeMethodologyLoader, getDefaultRuntimeLoader, resetDefaultRuntimeLoader, } from './runtime-methodology-loader.js';
// Export hot reload coordinator
export { MethodologyHotReloadCoordinator, createMethodologyHotReloadCoordinator, createMethodologyHotReloadRegistration, } from './methodology-hot-reload.js';
// Export shared Zod schema (SSOT for methodology validation)
export { MethodologySchema, MethodologyGateSchema, TemplateSuggestionSchema, validateMethodologySchema, } from './methodology-schema.js';
